lib/ddtrace/contrib/http/instrumentation.rb

```diff
@@ -45,14 +45,17 @@ def request(req, body = nil, &block) # :yield: +response+
                 if pin.tracer.enabled && !Datadog::Contrib::HTTP.should_skip_distributed_tracing?(pin)
                   Datadog::HTTPPropagator.inject!(span.context, req)
                 end
+
+                # Add additional request specific tags to the span.
+                annotate_span_with_req!(span, req)
               rescue StandardError => e
                 Datadog::Logger.log.error("error preparing span for http request: #{e}")
               ensure
                 response = super(req, body, &block)
               end
 
-              # Add additional tags to the span.
-              annotate_span!(span, req, response)
+              # Add additional response specific tags to the span.
+              annotate_span_with_res!(span, response)
```

Comment:
Hmmm, I think in the case where an exception is raised, it won't actually set the error on the request span. (If `super` raises an error, this annotation will be bypassed.)

It might be a good idea to leverage the `on_error` function as a part of `trace` to catch and set the error in this case. We should then also add a unit test for this scenario.

Reply:
pushed up an update that uses `on_error` and added a bit more to the specs to have some more unit tests for possible errors. What a thoughtful little hook, makes this much easier.

there's a bit of an open question on how to handle setting status codes in the case of a request that throws an exception, i suppose it makes sense to leave it nil if it isnt specified by the error (which it shouldnt be afaik) but wasnt sure if there was a best practice here.

Reply:
We have an example on how this is handled today in our Faraday integration: we provide all information we can, but we are forced to leave some of it blank, like the status code you mentioned.

Here's an example of a Faraday test to cover this same case: https://github.com/DataDog/dd-trace-rb/pull/909/files

Reply:
thanks @marcotc that makes sense. the update i believe matches that behavior so we provide what we can and leave the rest blank (not assuming defaults on things like status code).

oddly enough it looks like one of the tests flaked unrelated to this change for `test-2.4` ? 

```
1) Datadog::Contrib::Ethon with simple Easy & headers override behaves like instrumented request instrumented request created span response has not found status has tag with status code
     Failure/Error: expect(span.get_tag(Datadog::Ext::HTTP::STATUS_CODE)).to eq(status.to_f)
     
       expected: 404.0
            got: nil
     
       (compared using ==)
     Shared Example Group: "instrumented request" called from ./spec/ddtrace/contrib/ethon/integration_spec.rb:71
```

Reply:
Yeah this might be a flaky test. You can "rerun from failed" in CircleCI without having to repush.